### DISCUSSION TITLE
[Proposal]: Nullable reference type checks should not be performed on members of abstract class

### DISCUSSION THREAD

**Discussion by SetTrend** [dotnet/csharplang#4359] [Language Ideas]
# FEATURE_NAME

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary / Motivation
Currently, nullable reference checks are performed on non-private properties/fields of abstract classes.

![grafik](https://user-images.githubusercontent.com/9283914/105789538-ed816700-5f82-11eb-8017-5e15a562a4c3.png)

Abstract classes, however, cannot be instantiated. Instead, a derived class _must_ be implemented to get the abstract class to life.

So, I propose to omit corresponding nullable reference checks from the constructor of abstract classes and, instead, raise them to occur on constructors of derived, non-abstract, implementing classes:

![Nullable reference checks on abstract class](https://user-images.githubusercontent.com/9283914/105792152-6be00800-5f87-11eb-8ac5-c7f8eec0de5f.gif)

Currently, it is required to assign redundant dummy values to such kind of properties/fields, which is cumbersome depending on the type of object to create. It doesn't add meaning to the code and bloats the result.

**Comment by Phantonia** [comment]
I kind of see your point, but I disagree. In my opinion, each field should be `private`. Every class should ensure itself that instances are in a valid state. It's also quite telling that you can't even assign `readonly` field of the base class in a sub class's constructor. 
The better way would be assigning the field in the base class's constructor and making it a parameter.

**Reply by SetTrend** [reply]
Yes, i apprehend. The preference of fields vs. properties has been and is an ongoing discussion for decades. I prefer fields over properties because they yield a faster compile result.

In the end, if you're not using interfaces and if you need no logic behind getters/setters (that's what auto properties do), a field is perfectly sufficient.

**Reply by jnm2** [reply]
> I prefer fields over properties because they yield a faster compile result.

I don't believe this is usually true in practice. Jitting is smart enough to inline the field access of simple property accessors, making property access zero-cost compared to field access.

**Reply by theunrepentantgeek** [reply]
This is easy to test - here's a [SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgAYBYAKDIEYabiBmAAjNYGEIA7DbAJY8YUVjQDeNVtNYAHKAIBu2DDFaxsAE14AbAJ7s65VgH0e2fDADcUmbektWAZQwKeAc1YA5C2vGt3GAxWAF4APlNzSytWAF87ahkxRJlHYhRWABEYfAgACgBKe1ZJFKTpAHoK4joATjyzXwKbMvKa+p9LZuKk9sKWpNiaIeogA===) example.

Assume this class definition:

``` csharp
public class Container 
{
    private readonly string _name;    
    public string Name { get => _name; }   
}
```

Using the field directly, `Console.WriteLine(_name);` compiles to this in release mode:

``` x64
    L0004: mov rcx, [rcx+8]
    L0008: call System.Console.WriteLine(System.String)
```

Using the property, `Console.WriteLine(Name)` compiles to this:

``` x64
    L0004: mov rcx, [rcx+8]
    L0008: call System.Console.WriteLine(System.String)
```

Yup. This shows that trivial property setters are definitely elided in release builds.

**Reply by SetTrend** [reply]
Nonetheless:

```c#
public class Container 
{
    private int _x;
    private int _y { get; set; }
    
    public Container()
    {
        _x=1;
        _y=2;
    }
}
```

... yields (in full length):

```asm
Container..ctor()
    L0000: mov dword ptr [rcx+8], 1
    L0007: mov dword ptr [rcx+0xc], 2
    L000e: ret

Container.get__y()
    L0000: mov eax, [rcx+0xc]
    L0003: ret

Container.set__y(Int32)
    L0000: mov [rcx+0xc], edx
    L0003: ret
```

... which is 7 opcodes compared to 3 (a surplus of 150% percent).

It cannot be precluded that calls to `get__y()`/`set__y(Int32)` will be performed under certain circumstances. Frankly, the examples given are very trivial.

**Reply by theunrepentantgeek** [reply]
So?

A micro-optimization that saves a handful of opcodes _might_ (just _might_, if we ignore CPU caching) save a nanosecond or two.

You'd need an astonishingly large number of them to have any detectable effect - and as soon as you start interacting with anything out of thread (another thread, a system call, another process, network access to a database or another server) statistical variances in timing of those will completely dominate the benchmark results.

**Reply by huoyaoyuan** [reply]
> I prefer fields over properties because they yield a faster compile result.

Talking about "fast" at this level makes very wrong sense. This is a common mistake for ones casual about performance. **If non-virtual auto property produces worse code, it should be treated as a compiler bug.**

> ... which is 7 opcodes compared to 3 (a surplus of 150% percent).

Sharplab forces compilation of `get__y` and `set__y` to show them. In real world they will be always inlined and never compiled as separated methods.

Performance difference on these can be much smaller than you think. Don't let wrong sense of performance disturb the design.

**Reply by SetTrend** [reply]
@theunrepentantgeek : You seem to forget IoT devices. Not all of them allow to spend loads of memory unnecessarily. And you seem to ignore repetitive tasks, like loops.

Services are permanently running loops all the time, thereby burdening machine resources. In contrast to your assumption, hundreds of software packages are permanently competing for resources on a machine.

@huoyaoyuan : Thank you for your valuable comment. I'm going to check this later.

**Comment by huoyaoyuan** [comment]
This is covered by required properties. The abstract class will be able to require derived class to initialize the member, and it can be safely treated non-null.

**Reply by SetTrend** [reply]
Perhaps there's a misunderstanding ... If I change the field into a property, nothing changes:

![grafik](https://user-images.githubusercontent.com/9283914/105825902-6b179800-5fc0-11eb-9331-14aca3c9c4f0.png)

Would you mind elaborating?

**Reply by huoyaoyuan** [reply]
Required property is on the working set. See #4209 .
In the desired approach, there will be an error on constructor of `Derived` that it doesn't initialize `Prop`.

**Reply by SetTrend** [reply]
Ah, interesting. I see ...

Yet, for _abstract_ classes, I don't believe additional explicit adornment to be necessary as all is clear from the semantics alone.

With abstract classes it's obvious from their semantics that they are only a skeleton, a bare scaffold, unable to live on its own, requiring an aditional class to give meaning to the scaffold in coalescence. Performing selfcontained checks on an "unfertilized entity" doesn't make sense to me.

Moreover, restricting to properties alone is a biased restriction, based on fashion and dogma, rather than semantics. I tend to believe that compilers as well as languages should not restrict semantics by inclination. Instead they should particularly support alternative, yet valid, approaches, giving room to other ideas and possible improvement.

**Reply by huoyaoyuan** [reply]
You should use abstract properties in this case. The derived class has full control and responsibility of the property.
Abstract class can have non-abstract part.

**Reply by YairHalberstadt** [reply]
I think it quite likely that `required` would be allowed for both properties and fields.

Making a field implicitly required for an abstract class means that adding/removing an initializer from an abstract class changes its public contract. C# is of the opinion that all public contracts should be explicit.

**Reply by SetTrend** [reply]
I verified now that indeed automatic properties and fields are basically the same.

So, if they are practically the same, why not be consistent and finally remove the formal distinction between Fields and Properties? It doesn't make sense to formally distinguish between both if the is no difference in the end.

So, I propose to remove fields from the CLR and let the current field syntax be the new syntax for automated properties:

![Replace automatic properties with fields](https://user-images.githubusercontent.com/9283914/106143261-d2754980-6172-11eb-9c47-7c729b6811be.gif)

This would be a great relief. We finally could get rid of this tedious, mindless, repetitive writing of `{ get; set; }` all the time and just hit a simple `;` instead.

It really doesn't make sense to distinguish the indistinguishable.

**Reply by huoyaoyuan** [reply]
Auto properties can have different accessibilities on accessors.
Fields can be taken by `ref`.
They are never the same.

**Reply by SetTrend** [reply]
About the first issue you mentioned: That wouldn't be an issue. `accessor datatype name;` would just be a replacement for `accessor datatype name { get; set; }`. Anything else can still be written as usual (e.g. `accessor datatype name { get; private set}`).

For the second: In your previous post you mentioned there wouldn't be a difference. So, is there a difference now or not?

**Reply by huoyaoyuan** [reply]
There won't be a performance difference. There are usage differences.

**Reply by huoyaoyuan** [reply]
Specifically saying, an non-virtual auto property with same accessibility (`{ get; set; }`) is strictly covered by a field in non-reflection code, but it gives you the right to change the getter or setter body in the future.

I find myself seldom writing `public ... { get {} set {} }`. I mostly write init-only properties, and properties does action on set.

**Reply by SetTrend** [reply]
If you'd want to be explicit on `ref`s there'd be an option to explicitly allow refs on properties. In the end, a `ref` to a property would just be a collection of an `Action` and a `Func`.

**Reply by SetTrend** [reply]
> I find myself seldom writing public ... { get {} set {} }. 

Microsoft Entity Framework makes heavy use of so-called "POCO"s.

**Reply by huoyaoyuan** [reply]
Using `Action` or `Func` to replace `ref` hurts a lot of performance, or make some interlocked operations impossible. Plain `get` or `set` of auto property hurts 0.
That is off-topic and would be covered by `ref` accessor: `{ get; set; ref; }`.

**Reply by HaloFour** [reply]
> It really doesn't make sense to distinguish the indistinguishable.

They are distinguishable, even though the JIT can inline the accessor methods.  Those accessor methods could have different accessibility modifiers.  They could have attributes applied to them.  They can have modopts/modreqs applied to them.  They do prevent direct access to the field via IL.  You can't take a `ref` to the field directly.

The fact that the JIT can eliminate the runtime penalty of auto-properties (or simple properties in general) doesn't make them the same.

**Reply by SetTrend** [reply]
Let's focus: We're strictly talking about auto properties here, nothing else. Nothing with code behind. Auto properties with _no code_ are basically fields. Nothing else.

**Reply by SetTrend** [reply]
It's just a matter of declaration whether auto properties would allow `ref`, like @huoyaoyuan suggested.

**Reply by huoyaoyuan** [reply]
Future modification right without breaking change is often not considered in application, but very important in libraries. That's the key value.
In applications you are totally free to use fields when you want to.

**Reply by SetTrend** [reply]
But there's not point in using fields. You can just use properties everywhere.

**Reply by huoyaoyuan** [reply]
Fields are **implementation** and properties are **contract**. You can also not use properties at all and use `GetXXX` `SetXXX` methods everywhere.

**Reply by SetTrend** [reply]
My off-topic proposal is just to take the burdon of always being required to write `{ get; set; }` for a thing that's compiled into a field anyway.

**Reply by SetTrend** [reply]
A private property isn't a **contract**, it's implementation.

**Reply by huoyaoyuan** [reply]
I never see usage of private property in guideline-followed codebases.

**Reply by SetTrend** [reply]
So, you're saying "we never did that before" - and that this should be the reason to thwart the improvement?

**Reply by 333fred** [reply]
Fwiw, Kotlin actually did this: there are no fields. Everything is a property. However, it would be a massive breaking change for C# to change that now.

**Comment by huoyaoyuan** [comment]
I'm going to terminate the off-topic and focus on the original question.

Nullability and abstract are **contract**, while fields are **implementation**.

> Abstract classes, however, cannot be instantiated. Instead, a derived class must be implemented to get the abstract class to life.

Yes. It must be derived. However, there's no forcing about **the derived class must initialize the non-null fields**. *Current contract* is **the base class constructor must initialize everything properly**. Modifying the contract will require additional annotations, which is *required fields/properties*.
The only way currently to force derived class to pass initialization value is via constructor.

**Reply by SetTrend** [reply]
Yes, I created a separate proposal for the off-topic as I share your opinion.

> However, there's no forcing about the derived class must initialize the non-null fields.

That's the current design flaw. And that's the core of my suggestion: It must be forced that a class deriving from an abstract class will make sure all properties not initialized by the abstract base class will be initialized.

**Reply by huoyaoyuan** [reply]
> not initialized by the abstract base

Derived class can't see the implementation code in base class (for cross assembly scenario), so there must be a way to annotate that, which is *required fields/properties*.

**Reply by SetTrend** [reply]
Absolutely, we don't disagree here. As I wrote before, your proposal is great, and I agree with it.

It just should be automatic, not explicit. There's no need to manually annotate the inevitable. The compiler may annotate the abstract class's non-private properties (fields) automatically when they are not initialized by the abstract class itself.

**Reply by huoyaoyuan** [reply]
> It just should be automatic, not explicit.

2 points:
There must be a way to do it. The language team is just starting the designing phase.
The general design flavor of C# is to avoid inferring public contract from implementation to prevent accidental break. So there could probably be an IDE helper to quickly do the annotation.

**Reply by SetTrend** [reply]
Sure. Sounds great to me.
